ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,407 @@
+
+// Create thread reducing stack usage to a minimum to reduce memory usage.
+// Returns error number (like pthread_create).
+static int create_thread_minstack(pthread_t *th, void *(*proc)(void *), void *arg);
+
+static inline void
+reap_pid(pid_t pid)
+{
+    int status;
+    while (waitpid(pid, &status, 0) < 0 && errno == EINTR)
+        continue;
+}
+
+static void *
+thread_proc_reap(void *arg)
+{
+    pid_t pid = (pid_t) (intptr_t) arg;
+
+    reap_pid(pid);
+
+    return NULL;
+}
+
+// Appends a string to *p_dest buffer.
+// It updates *p_dest to point after copied string.
+// Returns copied string.
+static char *
+append_string(char **p_dest, const char *s)
+{
+    char *const dest = *p_dest;
+    size_t const size = strlen(s) + 1;
+    memcpy(dest, s, size);
+    *p_dest = dest + size;
+    return dest;
+}
+
+static char**
+copy_string_list(value list)
+{
+    size_t strings_size = 0;
+    size_t list_size = 0;
+    char **res, **ptrs;
+    char *strings;
+
+    FOREACH_LIST(item, list) {
+        strings_size += strlen(String_val(Field(item, 0))) + 1;
+        ++list_size;
+    }
+
+    res = (char **) malloc(sizeof(char*) * (list_size + 1) + strings_size);
+    if (!res)
+        return NULL;
+
+    ptrs = res;
+    strings = (char *) (res + (list_size + 1));
+    FOREACH_LIST(item, list)
+        *ptrs++ = append_string(&strings, String_val(Field(item, 0)));
+    *ptrs = NULL;
+
+    return res;
+}
+
+static void
+close_fd(int *const p_fd)
+{
+    const int fd = *p_fd;
+    if (fd >= 0) {
+        *p_fd = -1;
+        close(fd);
+    }
+}
+
+typedef struct {
+    const char *err_msg;
+    pid_t pid;
+    msg_t msg;
+} safe_exec_result;
+
+static int
+safe_exec_with_helper(safe_exec_result *res, char **args, char **envs)
+{
+    int err = EINVAL;
+    char fd_string[48];
+    int pipe_fds[2] = { -1, -1 };
+
+    res->err_msg = "safe_exec";
+
+    if (!args[0] || !args[1] || !args[2])
+        return EINVAL;
+
+    if (strcmp(args[1], "-e") == 0) {
+        if (pipe(pipe_fds) < 0) {
+            res->err_msg = "pipe";
+            return errno;
+        }
+        sprintf(fd_string, "%d", pipe_fds[1]);
+        args[2] = fd_string;
+    }
+
+    sigset_t sigset, old_sigset;
+    int cancellation_state;
+
+    // Disable cancellation to avoid some signals.
+    // Glibc use some signals to handle thread cancellation.
+    pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cancellation_state);
+
+    // Block all possible signals to avoid receiving some in the child.
+    // Signal mask is inherited to new process/thread will start with
+    // all signals disabled and we can safely change them.
+    sigfillset(&sigset);
+    pthread_sigmask(SIG_BLOCK, &sigset, &old_sigset);
+
+    // fork
+    err = 0;
+    res->pid = vfork();
+    if (res->pid < 0) {
+        err = errno;
+    } else if (res->pid == 0) {
+        // child
+        if (pipe_fds[0] >= 0)
+            close(pipe_fds[0]);
+        execve(args[0], args, envs);
```

Comment:
This is now a lot closer to `vfork`, and should avoid previous concerns about what functions can be safely called in-between :)

(I haven't reviewed the rest yet)

Reply:
Trying to reduce that part (C code in Ocaml) the only type passed is a list of string (2 parameters) and the code doing the Ocaml -> C transformation (and vice versa) is in a separated function from the C function doing the job, as suggested by Ocaml manual.